src/combinatorics/ramsey/infinite.lean

```diff
@@ -0,0 +1,483 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jared Corduan.
+-/
+
+import logic.basic
+
+/-!
+# The Infinite Ramsey Theorem
+
+The main theorem, `infinite_ramsey_pairs_two_colors`, states that given
+a function from the unordered pairs of natural numbers to a set
+with two elements (often named `red` and `blue`), there exists an
+infinite set of natural numbers whose unordered pairs are all given
+the same color.
+
+The proof given here follows a very standard proof, such as the one
+given in chapter 1, theorem 5, of Graham, et al.
+
+The proof roughly goes as follows:
+
+Given a function `f` of unordered pairs into the colors `red` and `blue`,
+define a sequence `(Sᵢ, xᵢ)` so that:
+
+* (S₀, x₀) = (ℕ, 0)
+* Sᵢ₊₁ ⊆ Sᵢ
+* xᵢ₊₁ ∈ Sᵢ
+* xᵢ < xᵢ₊₁
+* f {xᵢ₊₁, y} = f {xᵢ₊₁, z} for all xᵢ₊₁ < y < z in Sₓ, x = xᵢ
+
+Then the set {xᵢ | i ∈ ℕ} has the property that
+f {xᵢ, xⱼ} = f {xᵢ, xₖ} for all i < j < k.
+If we color singletons in this set by f' xᵢ = f {xᵢ, xᵢ₊₁},
+then by the pigeon hole principle there is an infinite subset
+of the xᵢ's whose unordered pairs are all colored the same.
+
+## Notation
+
+This file uses `[S]²` to denote the set of unordered pairs of
+the given set `S`.
+
+## Implementation notes
+
+The set of two colors is defined as an emumeration.
+This definition is easy to work with, and easy to read,
+but comes at the cost of not generalizing.
+
+Unordered pairs are implemented as ordered pairs with the condition
+that the second element is strictly greater than the first.
+This definition is also easy to work with but does not generalize.
+
+## References
+
+*  [Graham, R.L. and Rothschild, B.L. and Spencer, J.H., *Ramsey Theory*][graham1990]
+-/
+
+open set
+open classical
+open nat
+
+section infinite_ramsey_pairs
+
+/--
+A set of natural numbers H is infinite if for any natural
+number there is a larger number in H.
+-/
+def infinite (H : set ℕ) := ∀ x : ℕ, ∃ y : ℕ, x < y ∧ y ∈ H
+
+/--
+There are two colors, red and blue.
+-/
+inductive color
+| red : color
+| blue : color
+
+/--
+An infinite set of natural numbers.
+-/
+structure Inf :=
+(s : set ℕ)
+(pf : infinite s)
+
+instance : has_coe Inf (set ℕ) := ⟨Inf.s⟩
+instance : has_mem ℕ Inf := ⟨λ n H, n ∈ H.s⟩
+instance : has_subset Inf := ⟨λ H₁ H₂, H₁.s ⊆ H₂.s⟩
+
+open color
+
+/--
+Given a function from natural numbers to colors,
+this property describes the numbers which are mapped to red.
+-/
+def reds (f : ℕ → color) (H : set ℕ) := (λ n : ℕ, f n = red ∧ n ∈ H )
+
+/--
+Given a function from natural numbers to colors,
+this property describes the numbers which are mapped to blues.
+-/
+def blues (f : ℕ → color) (H : set ℕ) := (λ n : ℕ, f n = blue ∧ H n)
+
+lemma lt_succ_sum : forall x y : ℕ, x < x + y + 1 :=
+λ x y : ℕ, lt_add_of_pos_right x (succ_pos y)
+
+/--
+Unordered pairs are defined as ordered pairs with the condition
+that the second element is strictly greater than the first.
+-/
+def unordered_pairs (S : set ℕ) :=
```

Comment:
This seems a bit adhoc. The more natural thing is probably `finset`s of cardinality 2. But those might be harder to use, because you cannot talk about `p.fst` and `p.snd`.

Reply:
yes, exactly. I tried to use finset, but I spent a lot of time fighting simple things. I tried this definition just to see if I could make faster progress, and indeed I could. But now that I have a proof it might make sense to go back and see if I can get finset to work.